### ISSUE TITLE
Invalid MEX-file error

### ISSUE THREAD

**Issue by Mathijssch** [casadi/casadi#2635]
### Version info: 
- CasADI v3.5.2 (latest) -- same problem occurs with v3.4.5
- MATLAB r2019b
- OS: Ubuntu 20.04

### Problem
I'm trying to test CasADI by creating a new symbol, and I get the following error message

    casadi.SX.sym('x');
       >> Invalid MEX-file '/home/mathijs/Documents/MATLAB/casadi/casadiMEX.mexa64': 
           <path to casadi>/casadi/casadiMEX.mexa64: undefined symbol:
          _ZN6casadi14SerializerBase5resetEv
   
         Error in casadi.GenSX.sym (line 432)
            [varargout{1:nargout}] = casadiMEX(316, varargin{:});

Any ideas on what could be the cause of this?

**Comment by jgillis** [comment]
Did you clear any older casadis from the matlabpath and restart matlab?

**Comment by Mathijssch** [comment]
Yes, several times to be 100% sure. Another thing I noticed, although I'm not sure how relevant it is: using an older version of CasADI (v3.4.5), the error complains about a different undefined symbol: 

    >>(...) casadi/casadiMEX.mexa64: undefined symbol:
    _ZNK6casadi8Callback4evalEPPKdPPdPxS4_Pv

**Comment by jgillis** [comment]
Can you confirm the following:
`nm casadiMEX.mexa64  | grep _ZN6casadi14SerializerBase5resetEv` should yield `U _ZN6casadi14SerializerBase5resetEv`
`nm libcasadi.so  | grep _ZN6casadi14SerializerBase5resetEv` should yield `T _ZN6casadi14SerializerBase5resetEv`


To debug further, start Matlab from the command line with:
`LD_DEBUG=libs matlab -nodesktop`
Once you see the Matlab command prompt, do the casadi `addpath` and do `casadi.SX.sym('x')`.
Post the output that comes after that command.

This tells you exactly which libcasadi is being loaded.

**Comment by Mathijssch** [comment]
> Can you confirm the following:
> `nm casadiMEX.mexa64 | grep _ZN6casadi14SerializerBase5resetEv` should yield `U _ZN6casadi14SerializerBase5resetEv`
> `nm libcasadi.so | grep _ZN6casadi14SerializerBase5resetEv` should yield `T _ZN6casadi14SerializerBase5resetEv`

Yes, this is behaving as expected. 

> To debug further, start Matlab from the command line with:
> `LD_DEBUG=libs matlab -nodesktop`
> Once you see the Matlab command prompt, do the casadi `addpath` and do `casadi.SX.sym('x')`.
> Post the output that comes after that command.
> 
> This tells you exactly which libcasadi is being loaded.

```
This is quite strange: it seems that it is yielding the correct output now. 
>> casadi.SX.sym('x')
87136:	
     87136:	calling init: /usr/local/MATLAB/R2019b/bin/glnxa64/builtins/mex/mwlibmex_builtinimpl.so
     87136:	
     87136:	find library=libcasadi.so [0]; searching
     87136:	 search path=/home/mathijs/Documents/MATLAB/casadi_latest/tls/haswell/x86_64:/home/mathijs/Documents/MATLAB/casadi_latest/tls/haswell:/home/mathijs/Documents/MATLAB/casadi_latest/tls/x86_64:/home/mathijs/Documents/MATLAB/casadi_latest/tls:/home/mathijs/Documents/MATLAB/casadi_latest/haswell/x86_64:/home/mathijs/Documents/MATLAB/casadi_latest/haswell:/home/mathijs/Documents/MATLAB/casadi_latest/x86_64:/home/mathijs/Documents/MATLAB/casadi_latest		(RPATH from file /home/mathijs/Documents/MATLAB/casadi_latest/casadiMEX.mexa64)
     87136:	  trying file=/home/mathijs/Documents/MATLAB/casadi_latest/tls/haswell/x86_64/libcasadi.so
     87136:	  trying file=/home/mathijs/Documents/MATLAB/casadi_latest/tls/haswell/libcasadi.so
     87136:	  trying file=/home/mathijs/Documents/MATLAB/casadi_latest/tls/x86_64/libcasadi.so
     87136:	  trying file=/home/mathijs/Documents/MATLAB/casadi_latest/tls/libcasadi.so
     87136:	  trying file=/home/mathijs/Documents/MATLAB/casadi_latest/haswell/x86_64/libcasadi.so
     87136:	  trying file=/home/mathijs/Documents/MATLAB/casadi_latest/haswell/libcasadi.so
     87136:	  trying file=/home/mathijs/Documents/MATLAB/casadi_latest/x86_64/libcasadi.so
     87136:	  trying file=/home/mathijs/Documents/MATLAB/casadi_latest/libcasadi.so
     87136:	
     87136:	
     87136:	calling init: /home/mathijs/Documents/MATLAB/casadi_latest/libcasadi.so
     87136:	
     87136:	
     87136:	calling init: /home/mathijs/Documents/MATLAB/casadi_latest/casadiMEX.mexa64
     87136:	
     87136:	/home/mathijs/Documents/MATLAB/casadi_latest/casadiMEX.mexa64: error: symbol lookup error: undefined symbol: mexFunctionAdapter (fatal)
     87136:	/home/mathijs/Documents/MATLAB/casadi_latest/casadiMEX.mexa64: error: symbol lookup error: undefined symbol: mexfilerequiredapiversion (fatal)
     87136:	/home/mathijs/Documents/MATLAB/casadi_latest/casadiMEX.mexa64: error: symbol lookup error: undefined symbol: mexfilerequiredapiversion_ (fatal)

ans = 

     87136:	
     87136:	calling init: /usr/local/MATLAB/R2019b/bin/glnxa64/builtins/graphics/drawnow/mwlibmwdrawnow_builtinimpl.so
     87136:	
x
```
And indeed, when I run MATLAB in the terminal (i.e., with `-nodesktop`), it does seem to be working properly again:
```
>> addpath('casadi_latest/')
>> casadi.SX.sym('x')

ans = 

x
```
However, when I remove `-nodesktop` and repeat the same steps from the MATLAB command line within the GUI, I get the original error. In an attempt to debug this, I tried starting MATLAB _graphically_ using `LD_DEBUG=libs matlab`, but -- somehow -- this seems to get rid of the error too. I tried this a few times now and consistently: I get the error when I run MATLAB graphically without any options, but I don't whenever I run MATLAB without `nodesktop` or with `LD_DEBUG=libs`.

**Comment by jgillis** [comment]
LD_DEBUG affecting behaviour is very odd indeed.
Can you run graphically with LD_PRELOAD=/home/mathijs/Documents/MATLAB/casadi_latest/libcasadi.so

**Comment by jgillis** [comment]
Perhaps you have older casadi libs in a folder in your PATH env variable

**Comment by Mathijssch** [comment]
> LD_DEBUG affecting behaviour is very odd indeed.
> Can you run graphically with LD_PRELOAD=/home/mathijs/Documents/MATLAB/casadi_latest/libcasadi.so

This seems to solve it.

> Perhaps you have older casadi libs in a folder in your PATH env variable

It doesn't look like it at first sight:
```
>>$PATH
bash: /home/mathijs/.cargo/bin:/home/mathijs/miniconda2/bin:/home/mathijs/miniconda2/condabin:/home/mathijs/julia-1.1.0/bin:/home/mathijs/.scripts:/home/mathijs/gems/bin:/usr/local/texlive/2018/bin/x86_64-linux:/home/mathijs/.cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/opt/gurobi810/linux64/bin
```
I would also be surprised because the error started occuring before I installed the latest version. Rather, I updated casadi in an attempt to solve the issue. Before the error occured, however, I did start using casadi in Python (in a separate anaconda environment which is inactive when I launch MATLAB). Is it possible that this interferes with the MATLAB version somehow? 
In this case, the only relevant directory in my PATH would be `miniconda2/bin`, which doesn't seem to contain anything casadi-related: 
```
tree /home/mathijs/miniconda2/bin
/home/mathijs/miniconda2/bin
├── 2to3
├── activate
├── autopoint
├── captoinfo -> tic
├── chardetect
├── clear
├── conda
├── conda-env
├── cph
├── c_rehash
├── dbus-cleanup-sockets
├── dbus-daemon
├── dbus-launch
├── dbus-monitor
├── dbus-run-session
├── dbus-send
├── dbus-test-tool
├── dbus-update-activation-environment
├── dbus-uuidgen
├── deactivate
├── dumppdf.py
├── dumppdf.pyc
├── easy_install
├── envsubst
├── epylint
├── gapplication
├── gdbus
├── gdbus-codegen
├── gettext
├── gettextize
├── gettext.sh
├── gio
├── gio-querymodules
├── glib-compile-resources
├── glib-compile-schemas
├── glib-genmarshal
├── glib-gettextize
├── glib-mkenums
├── gobject-query
├── gresource
├── gsettings
├── gtester
├── gtester-report
├── iconv
├── idle
├── infocmp
├── infotocap -> tic
├── isort
├── latin2ascii.py
├── latin2ascii.pyc
├── msgattrib
├── msgcat
├── msgcmp
├── msgcomm
├── msgconv
├── msgen
├── msgexec
├── msgfilter
├── msgfmt
├── msggrep
├── msginit
├── msgmerge
├── msgunfmt
├── msguniq
├── ncurses6-config
├── ncursesw6-config
├── ngettext
├── openssl
├── pcre-config
├── pcregrep
├── pcretest
├── pdf2txt.py
├── pdf2txt.pyc
├── pip
├── pip2
├── pip2.7
├── pybabel
├── pydoc
├── pygmentize
├── pylint
├── pylint-gui
├── pyreverse
├── python -> python2.7
├── python2 -> python2.7
├── python2.7
├── python2.7-config
├── python2-config -> python2.7-config
├── python-config -> python2.7-config
├── recode-sr-latin
├── reset -> tset
├── rst2html4.py
├── rst2html4.pyc
├── rst2html5.py
├── rst2html5.pyc
├── rst2html.py
├── rst2html.pyc
├── rst2latex.py
├── rst2latex.pyc
├── rst2man.py
├── rst2man.pyc
├── rst2odt_prepstyles.py
├── rst2odt_prepstyles.pyc
├── rst2odt.py
├── rst2odt.pyc
├── rst2pseudoxml.py
├── rst2pseudoxml.pyc
├── rst2s5.py
├── rst2s5.pyc
├── rst2xetex.py
├── rst2xetex.pyc
├── rst2xml.py
├── rst2xml.pyc
├── rstpep2html.py
├── rstpep2html.pyc
├── smtpd.py
├── sphinx-apidoc
├── sphinx-autogen
├── sphinx-build
├── sphinx-quickstart
├── sqlite3
├── sqlite3_analyzer
├── symilar
├── tabs
├── tclsh -> tclsh8.6
├── tclsh8.6
├── tic
├── toe
├── tput
├── tqdm
├── tset
├── wheel
├── wish -> wish8.6
├── wish8.6
├── xgettext
└── xmlwf
```

**Comment by jgillis** [comment]
I suggest `sudo updatedb`, `locate libcasadi.so` to bolden your claims

**Comment by Mathijssch** [comment]
This outputs 
```
/home/mathijs/Documents/MATLAB/casadi_latest/libcasadi.so
/home/mathijs/miniconda2/envs/RiskAverse/lib/libcasadi.so
/home/mathijs/miniconda2/envs/RiskAverse/lib/libcasadi.so.3.6
/home/mathijs/miniconda2/envs/RiskAverse/lib/python3.7/site-packages/casadi/libcasadi.so
/home/mathijs/miniconda2/pkgs/casadi-3.5.1-py37h44c41f7_5/lib/libcasadi.so
/home/mathijs/miniconda2/pkgs/casadi-3.5.1-py37h44c41f7_5/lib/libcasadi.so.3.6
```
but none of these directories are in my `PATH`.

**Comment by jgillis** [comment]
Does the problem persist if you temporarily rename all of those expect for the first to make them invisible to the loader?

**Comment by Mathijssch** [comment]
That solves it.

**Comment by jgillis** [comment]
It'd be interesting then to trace back and check which one was causing this

**Comment by jgillis** [comment]
Also, was it really the LD_DEBUG that affected behavior, or the fact that matlab was not started from a start menu but rather from a terminal?

**Comment by jgillis** [comment]
Matlb has an increasing integration with Python. I wonder if some Python paths are implictly added to the loader by Matlab. In any case, LD_DEBUG is supposed to reveal that without side effects

**Comment by Mathijssch** [comment]
> Also, was it really the LD_DEBUG that affected behavior, or the fact that matlab was not started from a start menu but rather from a terminal?

I started MATLAB from a terminal in both cases (I just repeated this to confirm):
 - I start matlab graphically as `LD_DEBUG=libs /usr/local/MATLAB/R2019b/bin/matlab`: no error 
 - I start matlab graphically as `/usr/local/MATLAB/R2019b/bin/matlab`: error; 

> Matlb has an increasing integration with Python. I wonder if some Python paths are implictly added to the loader by Matlab. In any case, LD_DEBUG is supposed to reveal that without side effects

This indeed seems to be the problem. The Python interpreter was for some reason (that I haven't found yet) loaded on startup whenever I opened matlab graphically. The python interpreter in MATLAB was set to the interpreter corresponding to the conda environment that has casadi installed, which resulted in the clash. When I set MATLAB's python interpreter to the default, casadi-less version (i.e., `pyenv('Version', '/usr/bin/python')`), the error disappeared.
